### PR TITLE
For #1653, #1500, Support GB28181, another solution.

### PR DESCRIPTION
**Pull request by winlinvip**
For #1653 and #1500

### CODE REVIEW

File: trunk/src/app/srs_app_gb28181.cpp [hunk 1]
```diff
@@ -0,0 +1,1439 @@
...
+
+        if (!completed){
+            return err;
+        }
+
+        //process completed frame data
+        char *payload = cache_payload[cur_timestamp]->bytes();
+        int  payload_len = cache_payload[cur_timestamp]->length();
+      
+        on_ps_stream(payload, payload_len, cur_timestamp);
```
Comment (winlinvip) [inline trunk/src/app/srs_app_gb28181.cpp]: Should check for errors.

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_gb28181.cpp [hunk 2]
```diff
@@ -0,0 +1,1439 @@
...
+            audio_fw.write(next_ps_pack,  payload_len, NULL);          
+#endif
+            
+			next_ps_pack = next_ps_pack + payload_len;
+            complete_len = complete_len + (payload_len + 9 + pse_pack->stuffing_length);
+            incomplete_len = ps_size - complete_len;
+
+            if (audio_enable && audio_stream.length() && can_send_ps_av_packet()) {
+                if ((err = gb28181->on_rtp_audio(&audio_stream, audio_pts)) != srs_success) {
+                    srs_trace("process ps audio packet error %s", err);
```
Comment (winlinvip) [inline trunk/src/app/srs_app_gb28181.cpp]: %s err may have exceptions. The err object needs to be released.

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_gb28181.cpp [hunk 3]
```diff
@@ -0,0 +1,1439 @@
...
+            break;
+        }
+    }
+
+    if (complete_len != ps_size){
+         srs_trace("gb28181:  client_id %s decode ps packet error! ps_size=%d  complete=%d \n", 
+                     session_id.c_str(), ps_size, complete_len);
+    }else if (video_stream.length() && can_send_ps_av_packet()) {
+         if ((err = gb28181->on_rtp_video(&video_stream, video_pts, keyframe)) != srs_success) {
+            srs_trace("process ps video packet error");
```
Comment (winlinvip) [inline trunk/src/app/srs_app_gb28181.cpp]: The error message is stored in 'err' and not printed in the log.

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_gb28181.cpp [hunk 4]
```diff
@@ -0,0 +1,1439 @@
...
+
+        SrsGb28181Conn* conn = NULL;
+
+        if ((err = fetch_or_create(req, &conn)) != srs_success) {
+            srs_trace("gb28181: conn create faild:%s", req->uri.c_str());
+            return  srs_error_wrap(err, "conn create faild");;
+        }
+        srs_assert(conn != NULL);
+        
+        // if (conn->register_status == Srs28181Unkonw)
```
Comment (winlinvip) [inline trunk/src/app/srs_app_gb28181.cpp]: Remove useless code directly.

`TRANS_BY_GPT3`

File: trunk/src/protocol/srs_sip_stack.cpp [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+    << "o=" << req->sip_auth_id << " 0 0 IN IP4 " << req->host << SRS_RTSP_CRLF
+    << "s=Play" << SRS_RTSP_CRLF
+    << "c=IN IP4 " << req->host << SRS_RTSP_CRLF
+    << "t=0 0" << SRS_RTSP_CRLF
+    << "m=video " << port <<" RTP/AVP 96 97 98 99" << SRS_RTSP_CRLF
+    << "a=recvonly" << SRS_RTSP_CRLF
+    << "a=rtpmap:96 PS/90000" << SRS_RTSP_CRLF
+    << "a=rtpmap:97 MPEG4/90000" << SRS_RTSP_CRLF
+    << "a=rtpmap:98 H264/90000" << SRS_RTSP_CRLF
+    << "a=rtpmap:99 H265/90000" << SRS_RTSP_CRLF
```
Comment (wasphin) [inline trunk/src/protocol/srs_sip_stack.cpp]: Are there plans to support these streams in `SrsPsRtpConn` if they have not been handled yet? If there are no plans to support them, it is recommended to remove this part.

`TRANS_BY_GPT3`

Comment (xialixin) [reply trunk/src/protocol/srs_sip_stack.cpp]: Currently only supports PS. The new submission has already been annotated. Thank you for the suggestion.

`TRANS_BY_GPT3`

**Pull request merged**